File: c/problem_33.c
#include "problem_33.h"
#include <stdio.h>
#include <string.h>
#include <assert.h>

void runningMedian(const int *nums, int n, double *out) {
    int sorted[256];
    int sortedLen = 0;
    for (int k = 0; k < n; k++) {
        int v = nums[k];
        int pos = sortedLen;
        for (int i = 0; i < sortedLen; i++) {
            if (sorted[i] >= v) { pos = i; break; }
        }
        for (int i = sortedLen; i > pos; i--) sorted[i] = sorted[i-1];
        sorted[pos] = v;
        sortedLen++;
        if (sortedLen % 2 == 1)
            out[k] = sorted[sortedLen / 2];
        else
            out[k] = (sorted[sortedLen/2 - 1] + sorted[sortedLen/2]) / 2.0;
    }
}

void problem33(void) {
    printf("Problem 33:\n");
    int nums[] = {2,1,5,7,2,0,5};
    double out[7], exp[] = {2,1.5,2,3.5,2,2,2};
    runningMedian(nums, 7, out);
    for (int i = 0; i < 7; i++) assert(out[i] == exp[i]);
    printf("All Tests Passed\n");
}

File: lua/problem_33.lua
local function insortedInsert(sorted, v)
    local lo, hi = 1, #sorted + 1
    while lo < hi do
        local mid = math.floor((lo + hi) / 2)
        if sorted[mid] < v then lo = mid + 1 else hi = mid end
    end
    table.insert(sorted, lo, v)
end

local function runningMedian(nums)
    local result = {}
    local sorted = {}
    for _, v in ipairs(nums) do
        insortedInsert(sorted, v)
        local n = #sorted
        if n % 2 == 1 then
            result[#result + 1] = sorted[math.floor(n / 2) + 1]
        else
            result[#result + 1] = (sorted[n / 2] + sorted[n / 2 + 1]) / 2.0
        end
    end
    return result
end

local function tableEq(a, b)
    if #a ~= #b then return false end
    for i = 1, #a do if a[i] ~= b[i] then return false end end
    return true
end

local function problem33()
    print("Problem 33:")
    local got = runningMedian({2,1,5,7,2,0,5})
    local exp = {2, 1.5, 2, 3.5, 2, 2, 2}
    assert(tableEq(got, exp), "assertion failed")
    print("All Tests Passed")
end

return { problem33 = problem33 }
